c-dependencies/js-compute-runtime/builtins/console.cpp

```diff
@@ -1,26 +1,330 @@
 #include "console.h"
 
+JS::Result<std::string> ToSource(JSContext *cx, JS::HandleValue val,
+                                 JS::MutableHandleObjectVector visitedObjects);
+
+/**
+ * Turn a handle of a Promise into a string which represents the promise.
+ * - If the promise is pending this will return "Promise { <pending> }"
+ * - If the promise is rejected this will return "Promise { <rejected> (rejected-value)}"
+ *  where rejected-value would be the ToSource representation of the rejected value.
+ * - If the promise is resolved this will return "Promise { resolved-value}"
+ *  where resolved-value would be the ToSource representation of the resolved value.
+ */
+JS::Result<std::string> PromiseToSource(JSContext *cx, JS::HandleObject obj,
+                                        JS::MutableHandleObjectVector visitedObjects) {
+  std::string message = "Promise { ";
+  JS::PromiseState state = JS::GetPromiseState(obj);
+  switch (state) {
+  case JS::PromiseState::Pending: {
+    message += "<pending> }";
+    break;
+  }
+  case JS::PromiseState::Fulfilled: {
+    JS::RootedValue value(cx, JS::GetPromiseResult(obj));
+    auto value_source = ToSource(cx, value, visitedObjects);
+    if (value_source.isErr()) {
+      return JS::Result<std::string>(JS::Error());
+    }
```

Comment:
What do you think about using `MOZ_TRY_VAR` here instead?
```suggestion
    auto value_source = MOZ_TRY_VAR(ToSource(cx, value, visitedObjects));
```

Reply:
I first moved to MOZ_TRY_VAR but when I did the performance optimisation you suggested for appending to an already existing string, it involved changing these to MOZ_TRY instead of MOZ_TRY_VAR as the functions no longer return strings and instead append in-place